Gems/MachineLearning/Code/Source/Debug/MachineLearningDebugTrainingWindow.cpp

```diff
@@ -0,0 +1,378 @@
+            trainingInstance->m_trainingCycle.m_testData.LoadArchive(trainingInstance->m_testDataName, trainingInstance->m_testLabelName);
+        }
+    }
+
+    void LogVectorN(const AZ::VectorN& value, const char* label)
+    {
+        AZStd::string vectorString(label);
+        for (AZStd::size_t iter = 0; iter < value.GetDimensionality(); ++iter)
+        {
+            vectorString += AZStd::string::format(" %.02f", value.GetElement(iter));
+        }
+        AZLOG_INFO(vectorString.c_str());
+    }
+
+    void MachineLearningDebugTrainingWindow::RecalculateAccuracy(TrainingInstance* trainingInstance, ILabeledTrainingData& data)
+    {
+        trainingInstance->m_trainingCycle.InitializeContexts();
+        trainingInstance->m_totalSamples = static_cast<int32_t>(data.GetSampleCount());
+        trainingInstance->m_correctPredictions = 0;
+        trainingInstance->m_incorrectPredictions = 0;
+        for (int32_t iter = 0; iter < trainingInstance->m_totalSamples; ++iter)
+        {
+            const AZ::VectorN& activations = data.GetDataByIndex(iter);
+            const AZ::VectorN& label = data.GetLabelByIndex(iter);
+            const AZ::VectorN* output = m_selectedModel->Forward(trainingInstance->m_trainingCycle.m_inferenceContext.get(), activations);
+            AZStd::size_t prediction = ArgMaxDecode(*output);
+            AZStd::size_t actual = ArgMaxDecode(label);
+            if (prediction == actual)
+            {
+                ++trainingInstance->m_correctPredictions;
+            }
+            else
+            {
+                ++trainingInstance->m_incorrectPredictions;
+            }
+
+            if (ml_logAccuracyValues)
+            {
+                LogVectorN(label, "Actual");
+                LogVectorN(*output, "Output");
+            }
+        }
+    }
+
+    void MachineLearningDebugTrainingWindow::DrawLayerParameters(TrainingInstance* trainingInstance, AZStd::size_t layerIndex)
+    {
+        if (trainingInstance->m_layerWeights.size() < layerIndex)
+        {
+            trainingInstance->m_layerWeights.resize(layerIndex + 1);
+            trainingInstance->m_layerWeights[layerIndex].Init("Weights", 250, ImGui::LYImGuiUtils::HistogramContainer::ViewType::Histogram, true, 0.0f, 1.0f, ImGui::LYImGuiUtils::HistogramContainer::AutoExpand);
+        }
+
+        if (trainingInstance->m_layerBiases.size() < layerIndex)
+        {
+            trainingInstance->m_layerBiases.resize(layerIndex + 1);
+            trainingInstance->m_layerBiases[layerIndex].Init("Biases", 250, ImGui::LYImGuiUtils::HistogramContainer::ViewType::Histogram, true, 0.0f, 1.0f, ImGui::LYImGuiUtils::HistogramContainer::AutoExpand);
+        }
+
+        //m_selectedModel->GetLayerWeights(layerIter), m_selectedModel->GetLayerBiases(layerIter)
+
+        //trainingInstance->m_layerWeights[layerIndex].Draw(ImGui::GetColumnWidth(), 200.0f);
+        //trainingInstance->m_layerBiases[layerIndex].Draw(ImGui::GetColumnWidth(), 200.0f);
```

Comment:
Commented-out code should get removed

Reply:
Yeah, to-be-implemented. Monitoring parameters and gradients is super essential to being able to debug and train a model successfully since there are so many parameters that need to be tuned to get proper convergence.